File: scripts/constructZipper.py
import os
import zipfile
import re
from pathlib import Path
from collections import defaultdict
from concurrent.futures import ThreadPoolExecutor, as_completed

def parse_filename(filename):
    """
    Parse a filename to extract the components needed for organization.
    Expected format: prefix-action_rotation-sequence.extension
    Example: woman_shadow-powerUp_-67.5-014.webp
    Returns: (prefix, action, rotation_part, full_stem)
    """
    # Remove extension
    stem = Path(filename).stem
    
    # Pattern to match: prefix-action_rotation-sequence
    # This handles cases like: woman_shadow-powerUp_-67.5-014
    pattern = r'^([^-]+)-([^_]+)_(.+)-\d+$'
    match = re.match(pattern, stem)
    
    if match:
        prefix = match.group(1)  # woman_shadow
        action = match.group(2)  # powerUp
        rotation_part = match.group(3)  # -67.5
        
        # Create the base name without sequence number
        base_name = f"{prefix}-{action}_{rotation_part}"
        
        return prefix, action, base_name
    else:
        # Fallback for files that don't match expected pattern
        parts = stem.split('-')
        if len(parts) >= 2:
            prefix = parts[0]
            action = parts[1].split('_')[0] if '_' in parts[1] else parts[1]
            return prefix, action, stem
        else:
            return stem, "unknown", stem

def create_zip_archive(args):
    """
    Create a single zip archive from a group of files.
    This function is designed to be run in a separate thread.
    """
    base_name, files, output_path = args
    
    if not files:
        return None
        
    # Use the first file's metadata for folder structure
    first_file = files[0]
    prefix = first_file['prefix']
    action = first_file['action']
    
    # Create directory structure: ConstructZips/prefix/action/
    target_dir = output_path / prefix / action
    target_dir.mkdir(parents=True, exist_ok=True)
    
    # Create zip file: action_rotation.zip (without prefix)
    # Extract action and rotation from base_name (format: prefix-action_rotation)
    action_rotation = base_name.split('-', 1)[1] if '-' in base_name else base_name
    zip_path = target_dir / f"{action_rotation}.zip"
    
    try:
        # Create or update the zip file
        with zipfile.ZipFile(zip_path, 'w', zipfile.ZIP_DEFLATED) as zip_file:
            for file_info in files:
                file_path = file_info['file_path']
                # Add file to zip with just its filename (not full path)
                zip_file.write(file_path, file_path.name)
        
        return f"Created {zip_path} with {len(files)} files"
    except Exception as e:
        return f"Error creating {zip_path}: {str(e)}"

def organize_files(source_dir, output_base_dir, max_workers=None):
    """
    Organize files from source_dir into folders and zip archives in output_base_dir.
    Uses multithreading to speed up the zip creation process.
    
    Args:
        source_dir: Source directory containing files to organize
        output_base_dir: Base directory for output
        max_workers: Maximum number of threads to use (None = auto-detect)
    """
    source_path = Path(source_dir)
    output_path = Path(output_base_dir) / "ConstructZips"
    
    if not source_path.exists():
        print(f"Source directory does not exist: {source_dir}")
        return
    
    # Group files by their base name (without sequence number)
    file_groups = defaultdict(list)
    
    print("Scanning files...")
    # Scan all files in the source directory
    for file_path in source_path.iterdir():
        if file_path.is_file():
            prefix, action, base_name = parse_filename(file_path.name)
            file_groups[base_name].append({
                'file_path': file_path,
                'prefix': prefix,
                'action': action,
                'base_name': base_name
            })
    
    print(f"Found {len(file_groups)} unique file groups")
    
    # Prepare arguments for multithreaded processing
    zip_tasks = []
    for base_name, files in file_groups.items():
        zip_tasks.append((base_name, files, output_path))
    
    # Use ThreadPoolExecutor for multithreaded processing
    if max_workers is None:
        max_workers = min(32, (os.cpu_count() or 1) + 4)  # Default ThreadPoolExecutor logic
    
    print(f"Processing {len(zip_tasks)} zip archives using {max_workers} threads...")
    
    completed_count = 0
    with ThreadPoolExecutor(max_workers=max_workers) as executor:
        # Submit all tasks
        future_to_task = {executor.submit(create_zip_archive, task): task for task in zip_tasks}
        
        # Process completed tasks
        for future in as_completed(future_to_task):
            result = future.result()
            if result:
                print(result)
                completed_count += 1
                if completed_count % 50 == 0:  # Progress update every 50 files
                    print(f"Progress: {completed_count}/{len(zip_tasks)} archives completed")
    
    print(f"Completed processing {completed_count} zip archives")

def main():
    """
    Main function to organize files.
    """
    # Get the current user's Downloads folder dynamically
    downloads_path = Path.home() / "Downloads"
    source_directory = downloads_path / "output"
    
    print("Starting file organization...")
    print(f"Source directory: {source_directory}")
    
    organize_files(source_directory, source_directory)
    
    print("File organization complete!")

if __name__ == "__main__":
    main()

File: scripts/constructZipper.js
"use strict";
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    var desc = Object.getOwnPropertyDescriptor(m, k);
    if (!desc || ("get" in desc ? !m.__esModule : desc.writable || desc.configurable)) {
      desc = { enumerable: true, get: function() { return m[k]; } };
    }
    Object.defineProperty(o, k2, desc);
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || (function () {
    var ownKeys = function(o) {
        ownKeys = Object.getOwnPropertyNames || function (o) {
            var ar = [];
            for (var k in o) if (Object.prototype.hasOwnProperty.call(o, k)) ar[ar.length] = k;
            return ar;
        };
        return ownKeys(o);
    };
    return function (mod) {
        if (mod && mod.__esModule) return mod;
        var result = {};
        if (mod != null) for (var k = ownKeys(mod), i = 0; i < k.length; i++) if (k[i] !== "default") __createBinding(result, mod, k[i]);
        __setModuleDefault(result, mod);
        return result;
    };
})();
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.organizeFiles = organizeFiles;
exports.parseFilename = parseFilename;
exports.createZipArchive = createZipArchive;
const fs = __importStar(require("fs"));
const path = __importStar(require("path"));
const os = __importStar(require("os"));
const archiver_1 = __importDefault(require("archiver"));
// ============================================================================
// FILE PARSING
// ============================================================================
/**
 * Parse a filename to extract the components needed for organization.
 * Expected format: prefix-action_rotation-sequence.extension
 * Example: woman_shadow-powerUp_-67.5-014.webp
 * Returns: { prefix, action, baseName }
 */
function parseFilename(filename) {
    // Remove extension
    const stem = path.parse(filename).name;
    // Pattern to match: prefix-action_rotation-sequence
    // This handles cases like: woman_shadow-powerUp_-67.5-014
    const pattern = /^([^-]+)-([^_]+)_(.+)-\d+$/;
    const match = stem.match(pattern);
    if (match) {
        const prefix = match[1]; // woman_shadow
        const action = match[2]; // powerUp
        const rotationPart = match[3]; // -67.5
        // Create the base name without sequence number
        const baseName = `${prefix}-${action}_${rotationPart}`;
        return { prefix, action, baseName };
    }
    else {
        // Fallback for files that don't match expected pattern
        const parts = stem.split('-');
        if (parts.length >= 2) {
            const prefix = parts[0];
            const action = parts[1].includes('_') ? parts[1].split('_')[0] : parts[1];
            return { prefix, action, baseName: stem };
        }
        else {
            return { prefix: stem, action: 'unknown', baseName: stem };
        }
    }
}
// ============================================================================
// ZIP ARCHIVE CREATION
// ============================================================================
/**
 * Create a single zip archive from a group of files.
 * Returns a Promise that resolves when the zip is created.
 */
async function createZipArchive(baseName, files, outputPath) {
    if (!files || files.length === 0) {
        return null;
    }
    // Use the first file's metadata for folder structure
    const firstFile = files[0];
    const { prefix, action } = firstFile;
    // Create directory structure: ConstructZips/prefix/action/
    const targetDir = path.join(outputPath, prefix, action);
    fs.mkdirSync(targetDir, { recursive: true });
    // Create zip file: action_rotation.zip (without prefix)
    // Extract action and rotation from baseName (format: prefix-action_rotation)
    const actionRotation = baseName.includes('-') ? baseName.split('-').slice(1).join('-') : baseName;
    const zipPath = path.join(targetDir, `${actionRotation}.zip`);
    return new Promise((resolve, reject) => {
        // Create a write stream for the zip file
        const output = fs.createWriteStream(zipPath);
        const archive = (0, archiver_1.default)('zip', {
            store: true // No compression - maximum speed
        });
        output.on('close', () => {
            resolve(`Created ${zipPath} with ${files.length} files (${archive.pointer()} bytes)`);
        });
        archive.on('error', (err) => {
            reject(`Error creating ${zipPath}: ${err.message}`);
        });
        // Pipe archive data to the file
        archive.pipe(output);
        // Add files to the archive
        for (const fileInfo of files) {
            const filePath = fileInfo.filePath;
            const fileName = path.basename(filePath);
            archive.file(filePath, { name: fileName });
        }
        // Finalize the archive
        archive.finalize();
    });
}
// ============================================================================
// FILE ORGANIZATION
// ============================================================================
/**
 * Organize files from sourceDir into folders and zip archives in outputBaseDir.
 * Uses Promise.all with batching to process multiple zips concurrently.
 */
async function organizeFiles(sourceDir, outputBaseDir, maxConcurrent = null) {
    const sourcePath = path.resolve(sourceDir);
    const outputPath = path.join(path.resolve(outputBaseDir), 'ConstructZips');
    if (!fs.existsSync(sourcePath)) {
        throw new Error(`Source directory does not exist: ${sourceDir}`);
    }
    // Group files by their base name (without sequence number)
    const fileGroups = {};
    console.log('Scanning files...');
    // Scan all files in the source directory
    const entries = fs.readdirSync(sourcePath, { withFileTypes: true });
    for (const entry of entries) {
        if (entry.isFile()) {
            const filePath = path.join(sourcePath, entry.name);
            const { prefix, action, baseName } = parseFilename(entry.name);
            if (!fileGroups[baseName]) {
                fileGroups[baseName] = [];
            }
            fileGroups[baseName].push({
                filePath,
                prefix,
                action,
                baseName
            });
        }
    }
    const groupCount = Object.keys(fileGroups).length;
    console.log(`Found ${groupCount} unique file groups`);
    // Prepare tasks for processing
    const zipTasks = Object.entries(fileGroups).map(([baseName, files]) => ({
        baseName,
        files,
        outputPath
    }));
    // Determine maximum concurrent operations
    if (maxConcurrent === null) {
        maxConcurrent = Math.min(32, (os.cpus().length || 1) + 4);
    }
    console.log(`Processing ${zipTasks.length} zip archives using ${maxConcurrent} concurrent operations...`);
    // Process tasks in batches
    let completedCount = 0;
    const batchSize = maxConcurrent;
    for (let i = 0; i < zipTasks.length; i += batchSize) {
        const batch = zipTasks.slice(i, i + batchSize);
        const results = await Promise.allSettled(batch.map(task => createZipArchive(task.baseName, task.files, task.outputPath)));
        // Process results
        for (const result of results) {
            if (result.status === 'fulfilled' && result.value) {
                console.log(result.value);
                completedCount++;
                if (completedCount % 50 === 0) {
                    console.log(`Progress: ${completedCount}/${zipTasks.length} archives completed`);
                }
            }
            else if (result.status === 'rejected') {
                console.error(result.reason);
            }
        }
    }
    console.log(`Completed processing ${completedCount} zip archives`);
}
// ============================================================================
// MAIN
// ============================================================================
/**
 * Main function to organize files.
 */
async function main() {
    // Get source directory from command line argument or environment variable
    let sourceDirectory = process.argv[2] || process.env.OUTPUT_DIR;
    // Get destination directory from command line argument (if different from source)
    let destinationDirectory = process.argv[3];
    // Fallback to Downloads/output if no source directory specified
    if (!sourceDirectory) {
        const downloadsPath = path.join(os.homedir(), 'Downloads');
        sourceDirectory = path.join(downloadsPath, 'output');
    }
    // Resolve to absolute paths
    sourceDirectory = path.resolve(sourceDirectory);
    // If no destination specified, use source directory
    if (!destinationDirectory) {
        destinationDirectory = sourceDirectory;
    }
    else {
        destinationDirectory = path.resolve(destinationDirectory);
    }
    console.log('Starting file organization...');
    console.log(`Source directory: ${sourceDirectory}`);
    console.log(`Destination directory: ${destinationDirectory}`);
    try {
        await organizeFiles(sourceDirectory, destinationDirectory);
        console.log('File organization complete!');
    }
    catch (error) {
        console.error('Error during file organization:', error);
        process.exit(1);
    }
}
// Run if executed directly
if (require.main === module) {
    main().catch((error) => {
        console.error('Fatal error:', error);
        process.exit(1);
    });
}
